packages/finch-graphql/src/client/cache/QueryCache.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * QueryCache will be a client that will cache queries
+ * this makes it simple to be able to share information between queries without
+ * having to refetch information in other hooks.
+ *
+ * - We will need to serialize queries.
+ * - A way to turn on this, without it being on by default.
+ * - A way to notify hooks that something has updated.
+ */
+
+import { DocumentNode, print } from 'graphql';
+
+type Cache = Map<string, unknown>;
+
+export type Listener<Query extends unknown> = (updateInfo: Query) => void;
+
+interface ListenerMap {
+  [key: string]: Array<Listener<unknown>>;
+}
+
+interface QueryCacheOptions {
+  hydrate?: Cache;
+}
+
+const serializeQuery = (doc: DocumentNode, variables: any) => {
+  return `${print(doc)}:${JSON.stringify(variables)}`;
+};
+
+export class QueryCache {
+  cache: Cache;
+  listeners: ListenerMap;
+
+  constructor(options: QueryCacheOptions = {}) {
+    this.cache = options.hydrate ?? new Map();
+  }
+
+  subscribe<Query extends unknown>(
+    doc: DocumentNode,
+    variables: any,
+    listener: Listener<Query>,
+  ) {
+    const key = serializeQuery(doc, variables);
+    if (typeof this.listeners[key] === 'undefined') {
+      this.listeners[key] = [];
+    }
+    this.listeners[key].push(listener);
+    return () => {
+      const index = this.listeners[key].indexOf(listener);
+      this.listeners[key] = [
+        ...this.listeners[key].slice(0, index),
+        ...this.listeners[key].slice(index + 1),
+      ];
+    };
+  }
+
+  setQueryCache<Query extends unknown>(
```

Comment:
We should probably be consistent with public APIs. Both setters and getters should be `getCache` and `setCache`. A more general getter and setter can potentially allow for different caching system without a rewrite.